engine/src/xetex_font_info.rs

```diff
@@ -1877,48 +333,86 @@ unsafe extern "C" fn xbasename(mut name: *const libc::c_char) -> *const libc::c_
     }
     return base;
 }
+
 #[no_mangle]
-pub static mut gFreeTypeLibrary: FT_Library = 0 as *const FT_LibraryRec_ as FT_Library;
-static mut hbFontFuncs: *mut hb_font_funcs_t = 0 as *const hb_font_funcs_t as *mut hb_font_funcs_t;
-#[no_mangle]
-pub unsafe extern "C" fn XeTeXFontInst_base_ctor(
-    mut self_0: *mut XeTeXFontInst,
-    mut pathname: *const libc::c_char,
-    mut index: libc::c_int,
-    mut pointSize: libc::c_float,
-    mut status: *mut libc::c_int,
-) {
-    (*self_0).m_unitsPerEM = 0i32 as libc::c_ushort;
-    (*self_0).m_pointSize = pointSize;
-    (*self_0).m_ascent = 0i32 as libc::c_float;
-    (*self_0).m_descent = 0i32 as libc::c_float;
-    (*self_0).m_capHeight = 0i32 as libc::c_float;
-    (*self_0).m_xHeight = 0i32 as libc::c_float;
-    (*self_0).m_italicAngle = 0i32 as libc::c_float;
-    (*self_0).m_vertical = 0i32 != 0;
-    (*self_0).m_filename = 0 as *mut libc::c_char;
-    (*self_0).m_index = 0i32 as uint32_t;
-    (*self_0).m_ftFace = 0 as FT_Face;
-    (*self_0).m_backingData = 0 as *mut FT_Byte;
-    (*self_0).m_backingData2 = 0 as *mut FT_Byte;
-    (*self_0).m_hbFont = 0 as *mut hb_font_t;
-    (*self_0).m_subdtor = None;
-    if !pathname.is_null() {
-        XeTeXFontInst_initialize(self_0, pathname, index, status);
-    };
+pub static mut gFreeTypeLibrary: FT_Library = ptr::null_mut();
+static mut hbFontFuncs: *mut hb_font_funcs_t = ptr::null_mut();
+
+impl XeTeXFontInst {
+    pub unsafe fn new(
+        mut pathname: Option<&CStr>,
+        mut index: libc::c_int,
+        mut pointSize: libc::c_float,
+        mut status: *mut libc::c_int,
+    ) -> Self {
+        let mut neu = XeTeXFontInst {
+            m_unitsPerEM: 0,
+            m_pointSize: pointSize,
+            m_ascent: 0.0,
+            m_descent: 0.0,
+            m_capHeight: 0.0,
+            m_xHeight: 0.0,
+            m_italicAngle: 0.0,
+            m_vertical: false,
+            m_filename: None,
+            m_index: 0,
+            m_ftFace: std::ptr::null_mut(),
+            m_afm_backing_data: Vec::new(),
+            m_hbFont: std::ptr::null_mut(),
+            m_subdtor: None,
+            fk_font: None,
+        };
+        if let Some(path) = pathname {
+            if let Err(e) = neu.init(path, index) {
+                *status = e;
+            }
+        }
+        neu
+    }
+    pub unsafe fn init(
+        &mut self,
+        pathname: &CStr,
+        mut index: libc::c_int,
+    ) -> Result<(), i32> {
+        XeTeXFontInst_initialize(self, pathname, index)
+    }
+
+    pub unsafe extern "C" fn units_to_points(&self, units: f32) -> f32 {
+        (units * self.m_pointSize) / (self.m_unitsPerEM as f32)
+    }
+
+    pub unsafe fn get_glyph_bounds(&self, gid: GlyphID) -> GlyphBBox {
+        if let Some(font) = &self.fk_font {
+            if let Ok(rect) = font.typographic_bounds(gid as u32) {
```

Comment:
FYI, this is the only place we're actually querying font information so far. Further down, font-kit is also used to load the FT_Face.